Sources/SecureXPC/HandlerError.swift

```diff
@@ -0,0 +1,110 @@
+//
+//  HandlerError.swift
+//  SecureXPC
+//
+//  Created by Josh Kaplan on 2022-02-06
+//
+
+import Foundation
+
+/// Represents an error thrown by a server's handler when processing a client's request.
+public struct HandlerError: Error {
+    /// The localized description of the underlying error.
+    public let localizedDescription: String
+    /// The error thrown by the handler.
+    ///
+    /// This error will always be present on the server; however, it will only exist on the client if it could be encoded by the server and decoded by the client.
+    public let underlyingError: Error?
+    /// The name of the underlying error's type within the server's runtime.
+    ///
+    /// Used to disambiguate enum case name conflicts when decoding.
+    private let typeName: String
+    
+    private init(error: Error) {
+        self.underlyingError = error
+        self.localizedDescription = error.localizedDescription
+        self.typeName = String(describing: type(of: error))
+    }
+    
+    /// Wrap calls to an ``XPCServer``'s handlers in this.
+    static func rethrow<T>(_ handler: () throws -> T) throws -> T {
+        do {
+            return try handler()
+        } catch {
+            throw HandlerError(error: error)
+        }
+    }
+    
+    /// Wrap calls to an ``XPCServer``'s `async` handlers in this.
+    @available(macOS 10.15.0, *)
+    static func rethrow<T>(_ handler: () async throws -> T) async throws -> T {
+        do {
+            return try await handler()
+        } catch {
+            throw HandlerError(error: error)
+        }
+    }
+}
+
+extension HandlerError: Codable {
+    private enum CodingKeys: CodingKey {
+        case localizedDescription
+        case typeName
+        case underlyingError
+    }
+    
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encode(self.localizedDescription, forKey: .localizedDescription)
+        try container.encode(self.typeName, forKey: .typeName)
+        
+        if let underlyingError = underlyingError as? Codable {
+            // This is a bit hacky - we're "abusing" the ability to get a new encoder for a superclass and instead
+            // using it to have the error encode itself
```

Comment:
The goal is to get access to an `Encoder` instance so that we can pass it to the `underlyingError` which is a `Codable` and therefore has an `encode(...)` function. There's no way I know of to pass the `underlyingError` into one of the encoding containers' functions because while they take `Codable`, `Codable` itself does not conform to `Codable`. (As an aside, I find this aspect of Swift to be mildly infuriating.)

Do you know of/have ideas for a better way to do this? I'm all for making this less of a hack.